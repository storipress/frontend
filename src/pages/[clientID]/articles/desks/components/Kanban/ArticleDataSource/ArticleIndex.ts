import { Brand } from 'effect'
import type { SourceArticleIndex } from '.'

export type ArticleIndex = [stage: number, index: number] & Brand.Brand<'ArticleIndex'>

export interface Range {
  start: ArticleIndex
  end: ArticleIndex
}

const brand = Brand.refined<ArticleIndex>(
  (x) => Array.isArray(x) && x.length === 2 && x[0] >= 0 && x[1] >= 0,
  (x) => Brand.error(`Expected ${x} to be ArticleIndex`),
)

export function refine(x: Brand.Brand.Unbranded<ArticleIndex>): ArticleIndex {
  return brand(x)
}

export function make(stage: number, index: number): ArticleIndex {
  return brand([stage, index])
}

export function empty() {
  return make(0, 0)
}

export function is(x: Brand.Brand.Unbranded<ArticleIndex>): x is ArticleIndex {
  return brand.is(x)
}

export function increase(self: ArticleIndex): ArticleIndex {
  return add(self, 1)
}

export function add(self: ArticleIndex, n: number): ArticleIndex {
  return brand([self[0], self[1] + n])
}

export function fromSourceArticleIndex(
  sourceArticleIndex: SourceArticleIndex.SourceArticleIndex,
  perPage: number,
): ArticleIndex {
  return make(sourceArticleIndex[0], sourceArticleIndex[1] * perPage + sourceArticleIndex[2])
}
