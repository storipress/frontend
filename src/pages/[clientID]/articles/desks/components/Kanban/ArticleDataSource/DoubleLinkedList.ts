import type { Predicate } from 'effect'
import { Option } from 'effect'

export interface DoubleLinkedList {
  _previous: Option.Option<DoubleLinkedList>
  _next: Option.Option<DoubleLinkedList>
}

export interface Sibling {
  previous: Option.Option<DoubleLinkedList>
  next: Option.Option<DoubleLinkedList>
}

export function empty(): DoubleLinkedList {
  return { _previous: Option.none(), _next: Option.none() } as DoubleLinkedList
}

export function remove<Node extends DoubleLinkedList>(self: Node): Sibling {
  const previous = self._previous
  const next = self._next

  self._previous = Option.none()
  self._next = Option.none()

  if (Option.isSome(previous)) {
    previous.value._next = next
  }

  if (Option.isSome(next)) {
    next.value._previous = previous
  }

  return { previous, next }
}

export function insertBefore<Node extends DoubleLinkedList>(self: Node, other: Node): void {
  const previous = self._previous
  insertAfter(other, self)
  if (Option.isSome(previous)) {
    insertAfter(previous.value, other)
  }
}

export function insertAfter<Node extends DoubleLinkedList>(self: Node, other: Node): void {
  ensureLink(self, other)
}

export function replaceNode<Node extends DoubleLinkedList>(self: Node, other: Node): void {
  const { previous, next } = remove(self)
  if (Option.isSome(previous)) {
    ensureLink(previous.value, other)
  }
  if (Option.isSome(next)) {
    ensureLink(other, next.value)
  }
}

export function find<Node extends DoubleLinkedList, Out extends Node>(
  self: Node,
  predicate: Predicate.Refinement<Node, Out>,
): Option.Option<Out> {
  if (predicate(self)) {
    return Option.some(self)
  }

  let current = self
  while (Option.isSome(current._next)) {
    current = current._next.value as Node
    if (predicate(current)) {
      return Option.some(current)
    }
  }
  return Option.none()
}

function ensureLink<Node extends DoubleLinkedList>(self: Node, other: Node): void {
  self._next = Option.some(other)
  other._previous = Option.some(self)
}

export function head<Node extends DoubleLinkedList>(self: Node): Node {
  return Option.match(self._previous, {
    onSome: (previous) => head(previous as Node),
    onNone: () => self,
  })
}

export function toArray<Node extends DoubleLinkedList>(self: Node): ReadonlyArray<Node> {
  const first = head(self)
  return toArrayImpl(first) as ReadonlyArray<Node>
}

function toArrayImpl<Node extends DoubleLinkedList>(first: Node): ReadonlyArray<Node> {
  return Option.match(first._next, {
    onSome: (next) => [first, ...toArrayImpl(next as Node)],
    onNone: () => [first],
  })
}
