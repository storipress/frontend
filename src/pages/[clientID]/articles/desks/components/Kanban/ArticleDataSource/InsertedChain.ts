import type { Predicate } from 'effect'
import { Array, Chunk, HashMap, Option, Order, identity, pipe } from 'effect'
import invariant from 'tiny-invariant'
import * as DoubleLinkedList from './DoubleLinkedList'
import type * as ArticleIndex from './ArticleIndex'

interface BasicDocument {
  id: string
}

export interface PositionNode extends DoubleLinkedList.DoubleLinkedList {
  readonly _tag: 'PositionNode'
  pos: ArticleIndex.ArticleIndex
}

export interface DocumentNode<Document extends BasicDocument = BasicDocument>
  extends DoubleLinkedList.DoubleLinkedList {
  readonly _tag: 'DocumentNode'
  document: Document
}

export type Node<Document extends BasicDocument = BasicDocument> = PositionNode | DocumentNode<Document>

export interface WithPivot<Document extends BasicDocument = BasicDocument> {
  pivot: Option.Option<string>
  index: ArticleIndex.ArticleIndex
  document: Document
}

export interface WithOriginPosition<Document extends BasicDocument = BasicDocument> extends WithPivot<Document> {
  origin: ArticleIndex.ArticleIndex
}

export interface InsertedChain<Document extends BasicDocument = BasicDocument> {
  byId: HashMap.HashMap<string, DocumentNode<Document>>
  byStage: HashMap.HashMap<number, HashMap.HashMap<number, PositionNode>>
}

/**
 * Create empty InsertedChain
 * @returns InsertedChain
 */
export function empty<Document extends BasicDocument = BasicDocument>(): InsertedChain<Document> {
  return {
    byId: HashMap.empty<string, DocumentNode<Document>>(),
    byStage: HashMap.empty<number, HashMap.HashMap<number, PositionNode>>(),
  }
}

/**
 * Get article to be inserted at specific index
 * @param self self
 * @param index the index
 * @returns articles should be inserted before the specific index
 */
export function getChainByIndex<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain<Document>,
  index: ArticleIndex.ArticleIndex,
): Document[] {
  return pipe(
    self.byStage,
    HashMap.get(index[0]),
    Option.flatMap(HashMap.get(index[1])),
    Option.map(DoubleLinkedList.toArray<Node<Document>>),
    Option.getOrElse(() => [] as Node<Document>[]),
    Array.filter((node): node is DocumentNode<Document> => node._tag === 'DocumentNode'),
    Array.map((node) => node.document),
  )
}

/**
 * Remove specific article
 * @param self self
 * @param id article id
 * @returns InsertChain with article removed
 */
export function remove<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain<Document>,
  id: string,
): InsertedChain<Document> {
  return pipe(
    HashMap.get(self.byId, id),
    Option.map((node) => {
      DoubleLinkedList.remove(node)
      return {
        ...self,
        byId: HashMap.remove(self.byId, id),
      }
    }),
    Option.getOrElse(() => self),
  )
}

/**
 * Get all article that is after the specific index
 * @param self self
 * @param index the index
 * @returns Article that is after the specific index (inclusive)
 */
export function getRemainingAfter<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain,
  index: ArticleIndex.ArticleIndex,
): ReadonlyArray<Document> {
  return pipe(
    self.byStage,
    HashMap.get(index[0]),
    Option.map((stageMap) =>
      pipe(
        stageMap,
        HashMap.filter((_v, i) => i >= index[1]),
        HashMap.values,
        Chunk.fromIterable,
        Chunk.sortWith(
          identity,
          Order.mapInput(Order.number, (x) => x.pos[1]),
        ),
        Chunk.flatMap((item) =>
          pipe(
            item,
            DoubleLinkedList.toArray<Node<Document>>,
            Array.filter((node): node is DocumentNode<Document> => node._tag === 'DocumentNode'),
            Array.map((node) => node.document),
            Chunk.unsafeFromArray,
          ),
        ),
        Chunk.toArray,
      ),
    ),
    Option.getOrElse(() => [] as Document[]),
  )
}

/**
 * Filter articles with predicate
 * @param self self
 * @param predicate return false will remove article
 * @returns filtered InsertedChain
 */
export function filter<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain<Document>,
  predicate: Predicate.Predicate<Document>,
) {
  return pipe(
    self.byId,
    HashMap.values,
    Array.reduce(self, (chain, node) => {
      if (node._tag !== 'DocumentNode') {
        return chain
      }

      // It can stay
      if (predicate(node.document)) {
        return chain
      }

      return remove(chain, node.document.id)
    }),
  )
}

/**
 * Update specific node id
 * @param self self
 * @param oldId original id
 * @param newId new id
 * @returns updated result
 */
export function updateId<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain<Document>,
  oldId: string,
  newId: string,
): InsertedChain<Document> {
  const oldDocument = HashMap.get(self.byId, oldId)
  if (Option.isNone(oldDocument)) {
    return self
  }

  const position = getPosition(oldDocument.value)
  if (Option.isNone(position)) {
    return self
  }

  return pipe(remove(self, oldId), (chain) =>
    insertBefore(chain, {
      index: position.value,
      pivot: Option.none(),

      document: {
        ...oldDocument.value.document,

        id: newId,
      },
    }),
  )
}

/**
 * Modify article or remove it
 * @param self self
 * @param id article id
 * @param update return Option.none will remove article
 * @returns modified result
 */
export function modifyAt<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain<Document>,
  id: string,
  update: (document: Document) => Option.Option<Document>,
) {
  return pipe(
    HashMap.get(self.byId, id),
    Option.flatMap((node) => {
      const res = update(node.document)
      return Option.match(res, {
        onSome: (document) => {
          node.document = document
          return Option.some(self)
        },
        onNone: () => Option.some(remove(self, id)),
      })
    }),
    Option.getOrElse(() => self),
  )
}

/**
 * Insert article to InsertedChain, if there is pivot, it will use pivot point first, else it will fallback to index
 * @param self self
 * @param input insert position info
 * @returns updated InsertedChain
 */
export function insertBefore<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain<Document>,
  input: WithPivot<Document>,
) {
  const updated = remove(self, input.document.id)
  return pipe(
    tryInsertBeforeWithPivot(updated, input),
    Option.getOrElse(() => insertBeforeWithStage(updated, input)),
  )
}

/**
 * Insert article with specific pivot
 * User-facing code should use `insertBefore`
 *
 * @private
 */
export function tryInsertBeforeWithPivot<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain<Document>,
  { pivot, document }: Omit<WithPivot<Document>, 'index'>,
): Option.Option<InsertedChain<Document>> {
  return pipe(
    pivot,
    Option.flatMap((pivot) => HashMap.get(self.byId, pivot)),
    Option.map((pivotNode) => {
      const documentNode = makeDocumentNode(document)
      DoubleLinkedList.insertBefore(pivotNode, documentNode)
      return documentNode
    }),
    Option.map((documentNode) => ({
      ...self,
      byId: HashMap.set(self.byId, documentNode.document.id, documentNode),
    })),
  )
}

/**
 * Insert article to specific position, it's the fallback if `tryInsertBeforeWithPivot`
 * User-facing code should use `insertBefore`
 *
 * @private
 */
export function insertBeforeWithStage<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain<Document>,
  { index, document }: Omit<WithPivot<Document>, 'pivot'>,
): InsertedChain<Document> {
  const documentNode = makeDocumentNode(document)
  return {
    byId: HashMap.set(self.byId, documentNode.document.id, documentNode),
    byStage: HashMap.modifyAt(self.byStage, index[0], (stageMap) => {
      return pipe(
        stageMap,
        Option.getOrElse(() => HashMap.empty<number, PositionNode>()),
        HashMap.modifyAt(index[1], (positionNode) => {
          return pipe(
            Option.getOrElse(positionNode, () => makePositionNode(index)),
            (node) => {
              invariant(Option.isNone(node._next), 'position should be tail node')
              DoubleLinkedList.insertBefore<Node<Document>>(node, documentNode)
              return node
            },
            Option.some,
          )
        }),
        Option.some,
      )
    }),
  }
}

/**
 * Move article position
 * @param self self
 * @param param1 position info
 * @param param1.document document
 * @param param1.index document index
 * @param param1.origin document original index
 * @param param1.pivot document pivot
 * @returns updated result and restore function
 */
export function moveWithRestore<Document extends BasicDocument = BasicDocument>(
  self: InsertedChain<Document>,
  { document, index, origin, pivot }: WithOriginPosition<Document>,
): [chain: InsertedChain<Document>, restore: (updated: InsertedChain<Document>) => InsertedChain<Document>] {
  const restore = (updated: InsertedChain<Document>) => {
    return pipe(remove(updated, document.id), (chain) => insertBeforeWithStage(chain, { index: origin, document }))
  }

  return [insertBefore(self, { index, document, pivot }), restore]
}

export function getPosition<Document extends BasicDocument = BasicDocument>(
  documentNode: DocumentNode<Document>,
): Option.Option<ArticleIndex.ArticleIndex> {
  return pipe(
    DoubleLinkedList.find<Node, PositionNode>(
      documentNode,
      (node): node is PositionNode => node._tag === 'PositionNode',
    ),
    Option.map((node: PositionNode) => node.pos),
  )
}

/**
 * @private
 */
export function makeDocumentNode<Document extends BasicDocument = BasicDocument>(
  document: Document,
): DocumentNode<Document> {
  return {
    _tag: 'DocumentNode',
    document,
    ...DoubleLinkedList.empty(),
  }
}

/**
 * @private
 */
export function makePositionNode(index: ArticleIndex.ArticleIndex): PositionNode {
  return {
    _tag: 'PositionNode',
    pos: index,
    ...DoubleLinkedList.empty(),
  }
}
