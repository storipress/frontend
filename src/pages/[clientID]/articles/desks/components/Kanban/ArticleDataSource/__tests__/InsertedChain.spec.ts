import { describe, expect, it } from 'vitest'
import { Array, HashMap, Option, pipe } from 'effect'
import { omit } from 'lodash-es'
import * as InsertedChain from '../InsertedChain'
import * as ArticleIndex from '../ArticleIndex'
import * as DoubleLinkedList from '../DoubleLinkedList'

describe('empty', () => {
  it('should create an empty InsertedChain', () => {
    const chain = InsertedChain.empty()
    expect(HashMap.isEmpty(chain.byId)).toBe(true)
    expect(HashMap.isEmpty(chain.byStage)).toBe(true)
  })
})

describe('makeDocumentNode', () => {
  it('should create a DocumentNode with the correct structure', () => {
    const document = { id: 'doc1', title: 'Test Document' }
    const node = InsertedChain.makeDocumentNode(document)

    expect(node._tag).toBe('DocumentNode')
    expect(node.document).toEqual(document)
    expect(Option.isNone(node._previous)).toBe(true)
    expect(Option.isNone(node._next)).toBe(true)
  })
})

describe('makePositionNode', () => {
  it('should create a PositionNode with the correct structure', () => {
    const positionNode = InsertedChain.makePositionNode(ArticleIndex.refine([1, 1]))

    expect(positionNode._tag).toBe('PositionNode')
    expect(positionNode.pos).toEqual([1, 1])
    expect(Option.isNone(positionNode._previous)).toBe(true)
    expect(Option.isNone(positionNode._next)).toBe(true)
  })
})

describe('insertBeforeWithStage', () => {
  it('should insert a document node at the specified stage and position', () => {
    // Initial setup
    let chain = InsertedChain.empty()
    const document1 = { id: 'doc1', title: 'Document 1' }
    const document2 = { id: 'doc2', title: 'Document 2' }

    // Define positions
    const position1 = ArticleIndex.refine([1, 1]) // Stage 1, Position 1
    const position2 = ArticleIndex.refine([1, 2]) // Stage 1, Position 2

    // Insert the first document at position1
    chain = InsertedChain.insertBeforeWithStage(chain, { index: position1, document: document1 })

    // Validate the insertion of the first document
    const stage1Documents = pipe(
      chain.byStage,
      HashMap.get(1),
      Option.getOrElse(() => HashMap.empty<number, InsertedChain.PositionNode>()),
      HashMap.get(1),
      Option.map((x) => Array.make(x)),
      Option.getOrElse(() => []),
    )
    expect(stage1Documents).toHaveLength(1)
    expect(stage1Documents[0].pos).toEqual(position1)
    expect((Option.getOrThrow(stage1Documents[0]._previous) as InsertedChain.DocumentNode).document).toEqual(document1)

    // Insert the second document at position2
    chain = InsertedChain.insertBeforeWithStage(chain, { index: position2, document: document2 })

    // Validate the insertion of the second document
    const stage2Documents = pipe(
      chain.byStage,
      HashMap.get(1),
      Option.getOrElse(() => HashMap.empty<number, InsertedChain.PositionNode>()),
      HashMap.get(2),
      Option.map((x) => Array.make(x)),
      Option.getOrElse(() => []),
    )
    expect(stage2Documents).toHaveLength(1)
    expect(stage2Documents[0].pos).toEqual(position2)
    expect((Option.getOrThrow(stage2Documents[0]._previous) as InsertedChain.DocumentNode).document).toEqual(document2)

    expect(HashMap.size(chain.byStage)).toBe(1)
    expect(pipe(chain.byStage, HashMap.get(1), Option.getOrThrow, HashMap.size)).toBe(2)
    expect(HashMap.size(chain.byId)).toBe(2)

    expect(chain).toMatchSnapshot()
  })
})

describe('tryInsertBeforeWithPivot', () => {
  it('should insert a document node before a pivot node if the pivot exists', () => {
    // Initial setup
    const chain = InsertedChain.empty()
    const pivotDocument = { id: 'pivotDoc', title: 'Pivot Document' }
    const newDocument = { id: 'newDoc', title: 'New Document' }

    // Create and add pivot node to the chain
    const pivotNode = InsertedChain.makeDocumentNode(pivotDocument)
    chain.byId = HashMap.set(chain.byId, pivotDocument.id, pivotNode)

    // Attempt to insert the new document before the pivot
    const result = InsertedChain.tryInsertBeforeWithPivot(chain, {
      pivot: Option.some(pivotDocument.id),
      document: newDocument,
    })

    // Verify that the insertion was successful
    expect(Option.isSome(result)).toBe(true)
    const updatedChain = Option.getOrThrow(result)

    // Check that the new document node is correctly inserted before the pivot
    const insertedNode = Option.getOrThrow(HashMap.get(updatedChain.byId, newDocument.id))
    expect(DoubleLinkedList.toArray(insertedNode).map((x) => x.document)).toMatchInlineSnapshot(`
      [
        {
          "id": "newDoc",
          "title": "New Document",
        },
        {
          "id": "pivotDoc",
          "title": "Pivot Document",
        },
      ]
    `)
  })

  it('should return None if the pivot node does not exist', () => {
    // Initial setup with empty chain
    const chain = InsertedChain.empty()
    const document = { id: 'newDoc', title: 'New Document' }

    // Attempt to insert before a non-existent pivot
    const result = InsertedChain.tryInsertBeforeWithPivot(chain, { pivot: Option.some('nonExistentPivot'), document })

    // Verify that the insertion did not occur
    expect(Option.isNone(result)).toBe(true)
  })
})

describe('insertBefore', () => {
  it('should insert a document before a specified pivot if the pivot exists', () => {
    // Initial setup
    let chain = InsertedChain.empty()
    const pivotDocument = { id: 'pivotDoc', title: 'Pivot Document' }
    const newDocument = { id: 'newDoc', title: 'New Document' }

    // Create and add pivot node to the chain
    const pivotNode = InsertedChain.makeDocumentNode(pivotDocument)
    chain.byId = HashMap.set(chain.byId, pivotDocument.id, pivotNode)

    // Attempt to insert the new document before the pivot
    chain = InsertedChain.insertBefore(chain, {
      pivot: Option.some(pivotDocument.id),
      index: ArticleIndex.refine([1, 1]),
      document: newDocument,
    })

    // Verify the new document is correctly inserted before the pivot
    const insertedNode = Option.getOrThrow(HashMap.get(chain.byId, newDocument.id))

    expect(DoubleLinkedList.toArray(insertedNode).map((x) => x.document)).toMatchInlineSnapshot(`
      [
        {
          "id": "newDoc",
          "title": "New Document",
        },
        {
          "id": "pivotDoc",
          "title": "Pivot Document",
        },
      ]
    `)
  })

  it('should insert a document at a specific stage and position if pivot does not exist', () => {
    // Initial setup
    let chain = InsertedChain.empty()
    const newDocument = { id: 'newDoc', title: 'New Document' }
    const stagePosition = ArticleIndex.refine([1, 1]) // Example stage and position

    // Attempt to insert the new document at the specified stage and position
    chain = InsertedChain.insertBefore(chain, {
      index: stagePosition,
      document: newDocument,
      pivot: Option.some('nonExistentPivot'),
    })

    // Verify the new document is correctly inserted at the specified stage and position
    const insertedNode = Option.getOrThrow(HashMap.get(chain.byId, newDocument.id))

    expect(DoubleLinkedList.toArray(insertedNode).map((x) => omit(x, ['_previous', '_next']))).toMatchInlineSnapshot(`
      [
        {
          "_tag": "DocumentNode",
          "document": {
            "id": "newDoc",
            "title": "New Document",
          },
        },
        {
          "_tag": "PositionNode",
          "pos": [
            1,
            1,
          ],
        },
      ]
    `)
  })

  it('can handle multiple insert chain with pivot', () => {
    const documents = Array.makeBy(3, (i) => ({
      id: `doc${i + 1}`,
      title: `Document ${i + 1}`,
    }))

    let chain = InsertedChain.empty()
    for (let i = 0; i < documents.length; i += 1) {
      chain = InsertedChain.insertBefore(chain, {
        index: ArticleIndex.refine([1, 1]),
        document: documents[i],
        pivot: Option.fromNullable(documents[i - 1]?.id),
      })
    }

    expect(pipe(chain.byId, HashMap.size)).toBe(3)
    expect(pipe(chain.byStage, HashMap.size)).toBe(1)
    expect(pipe(chain.byStage, HashMap.get(1), Option.getOrThrow, HashMap.size)).toBe(1)
    expect(
      pipe(
        chain.byStage,
        HashMap.get(1),
        Option.getOrThrow,
        HashMap.get(1),
        Option.getOrThrow,
        DoubleLinkedList.toArray<InsertedChain.Node>,
        Array.map((x) => (x as InsertedChain.DocumentNode).document),
      ),
    ).toEqual([
      // we are using insertBefore, thus, the order is reversed
      ...Array.reverse(documents),
      // PositionNode
      undefined,
    ])
  })

  it('can handle multiple insert without pivot', () => {
    const documents = Array.makeBy(3, (i) => ({
      id: `doc${i + 1}`,
      title: `Document ${i + 1}`,
    }))

    const chain = Array.reduce(documents, InsertedChain.empty(), (chain, document) =>
      InsertedChain.insertBefore(chain, {
        index: ArticleIndex.refine([1, 1]),
        document,
        pivot: Option.none(),
      }),
    )

    expect(pipe(chain.byId, HashMap.size)).toBe(3)
    expect(pipe(chain.byStage, HashMap.size)).toBe(1)
    expect(pipe(chain.byStage, HashMap.get(1), Option.getOrThrow, HashMap.size)).toBe(1)
    expect(
      pipe(
        chain.byStage,
        HashMap.get(1),
        Option.getOrThrow,
        HashMap.get(1),
        Option.getOrThrow,
        DoubleLinkedList.toArray<InsertedChain.Node>,
        Array.map((x) => (x as InsertedChain.DocumentNode).document),
      ),
    ).toEqual([
      // expect to be same order as we keep insert at the same place
      ...documents,
      // PositionNode
      undefined,
    ])
  })

  it('should handle insert a document to different position', () => {
    let chain = InsertedChain.empty()
    chain = InsertedChain.insertBefore(chain, {
      document: {
        id: '1',
      },
      index: ArticleIndex.refine([0, 0]),
      pivot: Option.none(),
    })

    chain = InsertedChain.insertBefore(chain, {
      document: {
        id: '1',
      },
      index: ArticleIndex.refine([1, 0]),
      pivot: Option.none(),
    })

    const stage0Articles = pipe(
      chain.byStage,
      HashMap.get(0),
      Option.getOrThrow,
      HashMap.get(0),
      Option.getOrThrow,
      DoubleLinkedList.toArray<InsertedChain.Node>,
      Array.filter((x): x is InsertedChain.DocumentNode => x._tag === 'DocumentNode'),
    )

    expect(stage0Articles).toHaveLength(0)

    const stage1Articles = pipe(
      chain.byStage,
      HashMap.get(1),
      Option.getOrThrow,
      HashMap.get(0),
      Option.getOrThrow,
      DoubleLinkedList.toArray<InsertedChain.Node>,
      Array.filter((x): x is InsertedChain.DocumentNode => x._tag === 'DocumentNode'),
    )

    expect(stage1Articles).toHaveLength(1)
  })
})

describe('remove', () => {
  it('should remove document from chain', () => {
    let chain = InsertedChain.empty()
    chain = InsertedChain.insertBefore(chain, {
      document: {
        id: '1',
      },
      index: ArticleIndex.refine([0, 0]),
      pivot: Option.none(),
    })

    chain = InsertedChain.remove(chain, '1')

    expect(pipe(chain.byId, HashMap.size)).toBe(0)

    const stage0Articles = pipe(
      chain.byStage,
      HashMap.get(0),
      Option.getOrThrow,
      HashMap.get(0),
      Option.getOrThrow,
      DoubleLinkedList.toArray<InsertedChain.Node>,
      Array.filter((x): x is InsertedChain.DocumentNode => x._tag === 'DocumentNode'),
    )

    expect(stage0Articles).toHaveLength(0)
  })
})

describe('getRemainingAfter', () => {
  it('should return all article after specific position', () => {
    let chain = InsertedChain.empty()
    chain = InsertedChain.insertBefore(chain, {
      document: {
        id: '1',
      },
      index: ArticleIndex.refine([0, 0]),
      pivot: Option.none(),
    })

    chain = InsertedChain.insertBefore(chain, {
      document: {
        id: '2',
      },
      index: ArticleIndex.refine([0, 1]),
      pivot: Option.none(),
    })

    const remaining = InsertedChain.getRemainingAfter(chain, ArticleIndex.refine([0, 1]))

    expect(remaining).toEqual([
      {
        id: '2',
      },
    ])
  })
})
