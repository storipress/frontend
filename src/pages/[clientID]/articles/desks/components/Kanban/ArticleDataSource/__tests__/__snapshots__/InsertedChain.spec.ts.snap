// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`insertBeforeWithStage > should insert a document node at the specified stage and position 1`] = `
{
  "byId": {
    "_id": "HashMap",
    "values": [
      [
        "doc1",
        {
          "_next": {
            "_id": "Option",
            "_tag": "Some",
            "value": {
              "_next": {
                "_id": "Option",
                "_tag": "None",
              },
              "_previous": {
                "_id": "Option",
                "_tag": "Some",
                "value": [Circular],
              },
              "_tag": "PositionNode",
              "pos": [
                1,
                1,
              ],
            },
          },
          "_previous": {
            "_id": "Option",
            "_tag": "None",
          },
          "_tag": "DocumentNode",
          "document": {
            "id": "doc1",
            "title": "Document 1",
          },
        },
      ],
      [
        "doc2",
        {
          "_next": {
            "_id": "Option",
            "_tag": "Some",
            "value": {
              "_next": {
                "_id": "Option",
                "_tag": "None",
              },
              "_previous": {
                "_id": "Option",
                "_tag": "Some",
                "value": [Circular],
              },
              "_tag": "PositionNode",
              "pos": [
                1,
                2,
              ],
            },
          },
          "_previous": {
            "_id": "Option",
            "_tag": "None",
          },
          "_tag": "DocumentNode",
          "document": {
            "id": "doc2",
            "title": "Document 2",
          },
        },
      ],
    ],
  },
  "byStage": {
    "_id": "HashMap",
    "values": [
      [
        1,
        {
          "_id": "HashMap",
          "values": [
            [
              1,
              {
                "_next": {
                  "_id": "Option",
                  "_tag": "None",
                },
                "_previous": {
                  "_id": "Option",
                  "_tag": "Some",
                  "value": {
                    "_next": {
                      "_id": "Option",
                      "_tag": "Some",
                      "value": [Circular],
                    },
                    "_previous": {
                      "_id": "Option",
                      "_tag": "None",
                    },
                    "_tag": "DocumentNode",
                    "document": {
                      "id": "doc1",
                      "title": "Document 1",
                    },
                  },
                },
                "_tag": "PositionNode",
                "pos": [
                  1,
                  1,
                ],
              },
            ],
            [
              2,
              {
                "_next": {
                  "_id": "Option",
                  "_tag": "None",
                },
                "_previous": {
                  "_id": "Option",
                  "_tag": "Some",
                  "value": {
                    "_next": {
                      "_id": "Option",
                      "_tag": "Some",
                      "value": [Circular],
                    },
                    "_previous": {
                      "_id": "Option",
                      "_tag": "None",
                    },
                    "_tag": "DocumentNode",
                    "document": {
                      "id": "doc2",
                      "title": "Document 2",
                    },
                  },
                },
                "_tag": "PositionNode",
                "pos": [
                  1,
                  2,
                ],
              },
            ],
          ],
        },
      ],
    ],
  },
}
`;
